### PR TITLE
TIP Contact Us backend

### PR DESCRIPTION
**Pull request by hmzahmed**
- Create secure endpoint for submitting contact requests
- Create new table schema for tip contact requests
- Save information to our db
- Send slack message on each contact request

### CODE REVIEW

File: elixir/lib/sarkar/server/mis.ex [hunk 1]
```diff
@@ -88,6 +88,31 @@ defmodule EdMarkaz.Server.MIS do
 		end
 	end
 
+	# Make an post enpoint for sending data
+	post "/tip-contact-us" do
+		# Extracting Data from the connection Body and header
+		request_data = conn.body_params
+		# Check all three params school_id, client_id, token_id
+		[client_id] = get_req_header(conn, "client-id")
+		[ auth_token ] = get_req_header(conn, "auth-token")
```
Comment (mudssrali) [inline elixir/lib/sarkar/server/mis.ex:97]: spacing

File: elixir/lib/sarkar/server/mis.ex [hunk 2]
```diff
@@ -88,6 +88,31 @@ defmodule EdMarkaz.Server.MIS do
...
+
+		if String.length(client_id)>0 and String.length(auth_token)>0 do
```
Comment (mudssrali) [inline elixir/lib/sarkar/server/mis.ex:99]: I would say there would be two check here:
- is_binary (client-id)
- length (length of both client_id and auth_token is known to us)

File: elixir/lib/sarkar/server/mis.ex [hunk 3]
```diff
@@ -88,6 +88,31 @@ defmodule EdMarkaz.Server.MIS do
...
+			case EdMarkaz.TargetedInstructions.submit_contact_data(client_id, auth_token, request_data) do
+				{:success, resp} ->
+					EdMarkaz.Slack.send_alert("Received a Contact Request from #{request_data["request_person"]}", "#tip-contact-requests")
```
Comment (mudssrali) [inline elixir/lib/sarkar/server/mis.ex:102]: Create a complete request as a slack notification payload

File: elixir/lib/sarkar/server/mis.ex [hunk 4]
```diff
@@ -88,6 +88,31 @@ defmodule EdMarkaz.Server.MIS do
...
+					conn
+					|> send_resp(200, resp)
+				{:error, err, status} ->
+					conn
+					|> send_resp(status, err)
+			end
+		else
+				conn
+				|> send_resp(401, "client-id not submitted")
```
Comment (mudssrali) [inline elixir/lib/sarkar/server/mis.ex:111]: It's always good practice to return json, e.g. something like
```elixir
%{ status: 401, message: "Client id is invalid"}
```

File: elixir/lib/sarkar/targeted_instructions/targeted_instructions.ex [hunk 5]
```diff
@@ -122,4 +122,27 @@ defmodule EdMarkaz.TargetedInstructions do
...
+	@doc """
+	Documentation for submit_contact_data function
+	Requires 3 args, client_id, auth-token and a map with the data to submit
+	!!! IMPORTANT : THE MAP SHOULD ALWAYS INCLUDE THE school_id key value pair !!!
+	Checks in our tokens table if all the params school_id, client_id and auth_token match
+	Stores the contact request on our db if match is found
+	Returns with 401 error if not found
+	"""
+	@spec submit_contact_data(String.t(),String.t(),%{})::tuple()
+	def submit_contact_data(client_id,auth_token,contact_data) do
```
Comment (mudssrali) [inline elixir/lib/sarkar/targeted_instructions/targeted_instructions.ex:134]: A few things here:
- spacing
- save_contact_request

File: elixir/lib/sarkar/targeted_instructions/targeted_instructions.ex [hunk 6]
```diff
@@ -122,4 +122,27 @@ defmodule EdMarkaz.TargetedInstructions do
...
+		case EdMarkaz.DB.Postgres.query(EdMarkaz.DB,
+		"SELECT * from tokens where client_id=$1 and token=$2 and id=$3",
+		[client_id, auth_token, contact_data["school_id"]]) do
+			{:ok, %Postgrex.Result{num_rows: 0}} -> {:error, "Failed to Authorize", 401} # This happens when we cant find the client id in our db
```
Comment (mudssrali) [inline elixir/lib/sarkar/targeted_instructions/targeted_instructions.ex:138]: - comment before the code line
- `cant` -> cannot

File: elixir/lib/sarkar/targeted_instructions/targeted_instructions.ex [hunk 7]
```diff
@@ -122,4 +122,27 @@ defmodule EdMarkaz.TargetedInstructions do
...
+			{:ok, _resp} ->
+				EdMarkaz.DB.Postgres.query(EdMarkaz.DB,"INSERT INTO tip_contact_requests (value) VALUES ($1)",[contact_data]) # No issue found, proceed to make entry to database
+				{:success, "Submission Successful"}
+			{:error, err} -> # when something unknown happens
+				IO.inspect err
+				{:error, "An Unknown error has occured",500}
```
Comment (mudssrali) [inline elixir/lib/sarkar/targeted_instructions/targeted_instructions.ex:144]: - comments before each of code snippet not infront of
- spacing

**Review by mudssrali — CHANGES_REQUESTED** [review]
Thanks @hmzahmed that's awesome first backend PR 🎉 - a few comments

**Comment by hmzahmed** [comment]
Read the comments, will implement them first thing tomorrow

File: elixir/lib/sarkar/server/mis.ex (hunk 3 above)
Comment (hmzahmed) [reply elixir/lib/sarkar/server/mis.ex:102]: I dont understand what this means

**Pull request merged**